lib/elixir/lib/module/types/descr.ex

```diff
@@ -1278,8 +1278,130 @@ defmodule Module.Types.Descr do
 
   defp map_only?(descr), do: empty?(Map.delete(descr, :map))
 
-  # Union is list concatenation
-  defp map_union(dnf1, dnf2), do: dnf1 ++ (dnf2 -- dnf1)
+  defp map_union(dnf1, dnf2) do
+    # Union is just concatenation, but we rely on some optimization strategies to
+    # avoid the list to grow when possible
+
+    # first pass trying to identify patterns where two maps can be fused as one
+    with [{tag1, pos1, []}] <- dnf1,
+         [{tag2, pos2, []}] <- dnf2,
+         strategy when strategy != nil <- map_union_optimization_strategy(tag1, pos1, tag2, pos2) do
+      case strategy do
+        :all_equal ->
+          dnf1
+
+        :any_map ->
+          [{:open, %{}, []}]
+
+        {:one_key_difference, key, v1, v2} ->
+          new_pos = Map.put(pos1, key, union(v1, v2))
+          [{tag1, new_pos, []}]
+
+        :left_subtype_of_right ->
+          dnf2
+
+        :right_subtype_of_left ->
+          dnf1
+      end
```

Comment:
I believe we can encapsulate this and use it on `map_normalize` too?

Reply:
Oh, you mean instead of the `map_non_negated_fuse` / `fusible_maps` / `map_non_negated_fuse_pair` part right?
Seems it's doing very similar stuff, let's see!

Reply:
Yes, but I can also be the one giving that a try. We can merge this once the comments above are addressed, then I can work on the map_normalize and you work on tuples? :)

Reply:
I think I got it actually! Will push in a sec.

Reply:
[ec96574](https://github.com/elixir-lang/elixir/pull/14215/commits/ec965740d0b923d92a0ce4c489244680cc42c352)

I didn't spend time on renaming but I think we can probably use the term `fuse` / `fusion` etc more for consistency.
Feel free to rename as you see fit!